### DISCUSSION TITLE
How do you play a station from Sonos Radio?

### DISCUSSION THREAD

**Discussion by DennisGaida** [hklages/node-red-contrib-sonos-plus#245] [Q&A] [answered]
I know I can add a Sonos station to My Sonos, then use `mysonos.get.items` for the `title` and just use that to play via `group.play.mysonos`.

Is there any way to play a Sonos station by ID? Using `group.get.trackplus` I do see that Sonos Radio programs have a unique ID (with the service being ID 303. I know I can play TuneIn Radio stations, but when I enter a Sonos Radio ID it doesn't validate the radio regex.

**Comment by hklages** [comment]
The SONOS Radio implementation is strange - very often they still use TuneId IDs, but sometimes not. Lets see.  

Please add 2 SONOS radio stations to "My favorites" and do a "mysonos.export.item from My Sonos node" for both titles.  
Send this output to my email address "nrcsplus@gmail.com" or attach it as code to a comment here. 
Let me also know the Sonos Radio id for both.

I will compare the data and try to find a "template" where we only have to replace the id. That's how the tuneIn command works.

**Reply by DennisGaida** [reply]
I used `mysonos.get.items` instead of export since it gave more information. I added three stations and following is the output.

The three identifying information I could make out:
* `x-sonosapi-stream:tunein%3a4381?sid=303`
* `x-sonosapi-radio:154731?sid=303`
* `x-sonosapi-radio:154250?sid=303`

So it appears that even though it says Sonos Radio for the first radio station, Sonos might actually be using tunein? Or maybe they just call it tunein. The other two examples for radios are "music" radios or not traditional radios and they seem to be using the `x-sonosapi-radio` URN instead of `x-sonosapi-stream`.

For these three at least a regex could look like the following: `(x-sonosapi-radio|x-sonosapi-stream):(tunein%3a)?\d{4,6}\?sid=303` (https://regex101.com/r/2dg5Kn/1)

```json
{
   "id":"FV:2/28",
   "title":"NDR 90,3",
   "artist":"",
   "album":"",
   "description":"Sonos Radio Station",
   "uri":"x-sonosapi-stream:tunein%3a4381?sid=303&amp;flags=8232&amp;sn=7",
   "artUri":"https://sali.sonos.radio/image?w=60&amp;image=https%3A%2F%2Fcdn-profiles.tunein.com%2Fs2320%2Fimages%2Flogog.png%3Ft%3D154229&amp;partnerId=tunein",
   "metadata":"&lt;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quot;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:r=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&quot;&gt;&lt;item id=&quot;10092028tunein%3a4381&quot; parentID=&quot;00020000station:ndr 90,3&quot; restricted=&quot;true&quot;&gt;&lt;dc:title&gt;NDR 90,3&lt;/dc:title&gt;&lt;upnp:class&gt;object.item.audioItem.audioBroadcast&lt;/upnp:class&gt;&lt;desc id=&quot;cdudn&quot; nameSpace=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot;&gt;<anonymized>-Token&lt;/desc&gt;&lt;/item&gt;&lt;/DIDL-Lite&gt;",
   "sid":"303",
   "serviceName":"Sonos Radio",
   "upnpClass":"object.item.audioItem.audioBroadcast",
   "processingType":"stream"
}
```

```json
{
   "id":"FV:2/32",
   "title":"The Lighthouse",
   "artist":"",
   "album":"",
   "description":"Sonos Radio Station",
   "uri":"x-sonosapi-radio:154731?sid=303&amp;flags=0&amp;sn=7",
   "artUri":"",
   "metadata":"&lt;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quot;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:r=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&quot;&gt;&lt;item id=&quot;000c0000154731&quot; parentID=&quot;-1&quot; restricted=&quot;true&quot;&gt;&lt;dc:title&gt;The Lighthouse&lt;/dc:title&gt;&lt;upnp:class&gt;object.item.audioItem.audioBroadcast&lt;/upnp:class&gt;&lt;desc id=&quot;cdudn&quot; nameSpace=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot;&gt;<anonymized>-Token&lt;/desc&gt;&lt;/item&gt;&lt;/DIDL-Lite&gt;",
   "sid":"303",
   "serviceName":"Sonos Radio",
   "upnpClass":"object.item.audioItem.audioBroadcast",
   "processingType":"stream"
}
```

```json
{
   "id":"FV:2/33",
   "title":"Smooth Jazz",
   "artist":"",
   "album":"",
   "description":"Sonos Radio Station",
   "uri":"x-sonosapi-radio:154250?sid=303&amp;flags=0&amp;sn=7",
   "artUri":"",
   "metadata":"&lt;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quot;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:r=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&quot;&gt;&lt;item id=&quot;000c0000154250&quot; parentID=&quot;-1&quot; restricted=&quot;true&quot;&gt;&lt;dc:title&gt;Smooth Jazz&lt;/dc:title&gt;&lt;upnp:class&gt;object.item.audioItem.audioBroadcast&lt;/upnp:class&gt;&lt;desc id=&quot;cdudn&quot; nameSpace=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot;&gt;<anonymized>-Token&lt;/desc&gt;&lt;/item&gt;&lt;/DIDL-Lite&gt;",
   "sid":"303",
   "serviceName":"Sonos Radio",
   "upnpClass":"object.item.audioItem.audioBroadcast",
   "processingType":"stream"
}
```

**Comment by hklages** [comment, marked as answer]
Right - SONOS Radio often links to Tune In radios and the title does not give any hint what kind it is. That's makes it impossible to have one template. 

**For "pure" SONOS radio stations**
Are you an experienced Node-RED user? ( I am now in Asia and it is a bit difficult to access my SONOS). 
If yes you can try the following: 

1. Use the "The Lighthouse" for payload of "mysonos.export.item" and capture the result. Don't change the structure.
2. Edit the result in the following way
- in metadata: remove the title "The Lighthouse" and replace "154731" by "154250"
- in uri: replace  "154731" by "154250"
3. inject that result into "group.play.export 
This should play the other station "Smooth Jazz". 
For testing you can use the combination  mysonos.export.item then group.play.export. Thats the "Unversalwerkzeug". 
If you use an editor take care that there is really no other change like replacing line breaks or ".  

Does it work?
If yes you can use that as a template - similar to the Radio ID from TuneIn. Hopefully this id "154731" is always 6 digits.

**For the SONOS Radio "TuneIn" stations**
You can still use the group.play.tunein. 

**Other options**
Why do you dislike to add the stations to My Favorites? Do you intent to play tenth of different Radio Stations or do you get the "ID" (such as 154731) from a different system? 

There might be another general option. You can search any service (such as SONOS Radio) by title. So far I haven't implemented that.

**Reply by DennisGaida** [reply]
> If yes you can use that as a template

Yes, works!

* `msg.payload.uri` needs to be `x-sonosapi-radio:<ID>?sid=303&amp;flags=0&amp;sn=7` replacing the <ID>
* `msg.payload.metadata` needs to be `&lt;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quot;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:r=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/&quot;&gt;&lt;item id=&quot;000c0000<ID>&quot; parentID=&quot;-1&quot; restricted=&quot;true&quot;&gt;&lt;upnp:class&gt;object.item.audioItem.audioBroadcast&lt;/upnp:class&gt;&lt;desc id=&quot;cdudn&quot; nameSpace=&quot;urn:schemas-rinconnetworks-com:metadata-1-0/&quot;&gt;<TOKEN>-Token&lt;/desc&gt;&lt;/item&gt;&lt;/DIDL-Lite&gt;` replacing <ID> and the custom <TOKEN>

Now if only there was an easy way to figure out those IDs without adding to Favorites.

> Hopefully this id "154731" is always 6 digits.

My limited test of 3 Sonos Radio stations says yes.

> Why do you dislike to add the stations to My Favorites?

Clutter. Adding the station to favorites is not the worst option, but maybe I just want to play **this** station via Node-Red and never click on it via My Sonos / Favorites. Let's say I have multiple automations that play music depending on the time of day or something like that - I don't want all these stations in my favorites.

Ultimately it isn't that important - but we do have the `tunein` node, so maybe having a `sonosradio` node would be nice (or similarly `spotify` etc.) - if we were to figure out the rules for each service there could also some kind of general node: selecting the service from a dropdown (the service IDs are fixed) and entering a search string. But that's complicated for sure and something for the backlog. I'd love to search for e.g. Playlists (e.g. "Classical music for work") - Spotify or Amazon returns good results for that using the Sonos app. I wouldn't want to add all these playlists to My Sonos, but would like to able to play these randomly using automations.

**Comment by hklages** [comment]
Great. So you can work with that template. In case you have many stations you can store the ids in an array flow variable and reuse it - My Sonos is nothing then a list of objects of different types: stations, playlist, albums, ...)

I am not a friend of the group.play.tunein/group.queue.urispotify - I prefer the group.play.mysonos because it works for all services, playlists, albums, songs, etc. The spotify / tune in commands are "legacy" hacks.  

Next month I will try to implement the "title-find" for services - starting with "Sonos Radio". That could be a good addition to the group.play.mysonos and avoid cluttering My SONOS.

**Reply by DennisGaida** [reply]
Yeah tunein/uruspotify are very specific and a general thing would be way better.

I love you're trying to get to `title-find` that would be an awesome addition.
